### PR TITLE
New C# target for Haxe 5

### PR DESCRIPTION
**Pull request by jeremyfa**
Back in 2021 [I started a discussion](https://community.haxe.org/t/some-notes-feedback-about-c-target/3009/8) to share how much I didn't want the C# target to go away, despite all its flaws.

Time has passed and there have been some efforts with @SomeRanDev to create a new C# target [using Reflaxe](https://github.com/SomeRanDev/reflaxe.CSharp). It's been promising but more a proof of concept than anything really usable. Being busy on other matters that are more important that this, and knowing Haxe 4 C# target is still usable currently didn't help either.

Then, more time has passed and tools have evolved in a way that it became less scary, and more importantly less time consuming, to dig into Haxe compiler itself and its OCaml generators, and to start creating a whole new target from the grounds up, following the patterns of the existing targets and running it over the full test suite.

So that's what I did: I started creating a **new C# target for Haxe 5**, trying my best to follow the patterns of the existing targets (main reference being the JVM target, but also the previous Haxe 4 C# target, which does give a lot of C# specific insights).

I tried not to repeat the mistakes of the previous C# target: it is mostly self contained and it generates an intermediate C# AST (which follows an advice from @Simn I kept in mind when we were discussing about Reflaxe/CSharp which was doing "direct to text" transpilation at the time).

Although I can navigate through it and am familiar with pattern matching and manipulating ASTs, I am not particularly fluent in OCaml, but with the help of a coding assistant (Opus 4.5, mostly), it has been possible to iterate, take design decisions, write, review and test the code generation in a way that allows to converge to something that works. The `gencs.ml` file is quite large, which could be because it has a lot of comments (perks of using tooling that can write those easily, and makes the result more educational, for me at least), but maybe also because there are some redundancies, although I really tried to avoid them. C# being particularly picky regarding its type semantics could also be another reason a lot of patterns need to be handled, thus more code, but likely, there is room for improvements there as the priority has been correctness of the output and not necessarily keeping the OCaml concise.

Anyway, I am opening this Draft Pull Request today because I reached an important milestone: all `unit`, `sys` and `threads` tests are now passing correctly, so I believe it's a good time to start discussing about this, and question whether merging a new C# target to haxe could be considered, or if this new C# target is doomed to stay in a fork.

What is pretty clear on my end is that, merged or not, this work will be used for [Ceramic](https://ceramic-engine.com) to bring its Unity target to Haxe 5 (but also other projects I have that involve Haxe and C#), and any additional work needed to keep the target working with Haxe 5 in the future will be done : in other words I plan to maintain this as seriously as I maintain other projects like Ceramic in the forseeable future.

Some design choices on this target, that may be sometime very close, sometimes different from the previous one:

### Nullable types (`Null<T>`, optional params, in haxe) are transformed following those rules:

- Types that are inherently nullable (objects / reference types / strings ...) are not wrapped into anything, given that they are compatible with `null`

- Value/Primitive types are wrapped into [`haxe.lang.Null<T>`](https://github.com/jeremyfa/haxe/blob/new_csharp/std/cs/_cs/haxe/lang/Null.cs) (struct, on the stack, not the heap), and unwrapped as needed.

- C# generator takes care of generating proper coercion (using `.hasValue` instead of `!= null`, or calling proper methods to wrap/unwrap). The struct itself support some implicit conversions, although it is most of the time made explicit by the generator.

### Functions / Closures

- They are all subclasses of `haxe.lang.Function`

- Those functions can be called without boxing, thus without allocations on the heap (even for primitive types), because arguments are passed through a [`haxe.lang.Value`](https://github.com/jeremyfa/haxe/blob/new_csharp/std/cs/_cs/haxe/lang/Value.cs) struct (16 bytes) that can hold `bool`, `int`, `float`, `double`, `long`, any object including `null` or "no value", which also cover things like optional arguments.

- A caching mecanism is in place so that if you read `myObj.someMethod`, it will always return the same `haxe.lang.Function` instance, meaning you can use regular object equality to compare methods.

### AOT / NativeAOT

- The generated code is designed to be AOT compatible, even when using haxe reflection features, and is expected to work, on situations where C# reflection wouldn't be available. All the data needed to support haxe reflection is embedded in the related types. This could be fined-tuned in the future to allow disabling this additional data generation, if the code is expected to work in a scenario that does have full C# reflection available, but for now it is embedded by default, unless using `@:unreflective` meta (like in cpp target).

### Generics

- One of the trickiest parts of this target are generics, because Haxe generics are much more forgiving and flexible than C# generics. You can cast `SomeClass<Animal>` to `SomeClass<Dog>`, or even `SomeClass<Dynamic>` in Haxe, but this would just trigger a runtime error in C#.

- Haxe 4's C# target had a `-D erase-generics` define to remove generics. In this new C# target, this is somewhat the default: all haxe generics are erased and `object` is used instead, which is good to simplify the code generation and play better with AOT, as long as the generic type is a class type or a string, but value types in generics will be boxed (e.g. `MyType<Int>`). That being said, if a value type generic is needed, one can use `@:generic` to end up with `MyType_Int` with the proper type and no boxing, so given that Haxe provides this, we don't really need to rely on C# to do the same!

### Arrays

- Thanks to @tobil4sk for pointing me to [this issue](https://github.com/HaxeFoundation/haxe/issues/4872), which helped me decide how to implement arrays in the C# target. Basically it follows this locking idea from @ncannasse, where arrays are of "unknown" type at first, then get locked to a specific type if casted to an explicit `Array<Int>`, `Array<Float>` etc... The whole thing is implemented in a single `Array` class which switch storage as needed using native C# arrays (`int[]`, `double[]` etc...).

### Dynamic objects

- For now, dynamic objects, anonymous structures, rely on a `Dictionary<string,object>` lookup to resolve values for keys. Previous C# target was precomputing hash code to make the field access faster. There is no such thing on this new C# target at the moment, but I'll address this at some point.

### .NET externs

- At the moment, there is no such feature as `-net-lib someLib.dll`, but I created a generator, written in C#, that outputs Haxe externs for C#, from a DLL (api) + XML (docs) couple. The C# target comes by default with .NET Standard 2.1 API externs which have been generated this way. The externs are documented with proper doc comments fetched from the XML that comes with the DLL. There is no need to use a separate haxelib like `hxcs` or any dll /dotnet setup to export C# from this target. The same generator can be use to generate externs from any DLL, it's just that it is not handled by haxe.

- Not against supporting `-net-lib` later in a similar fashion as the previous C# target, but it is not a priority for me right now, and not gonna lie, it's actually nice to be able to navigate through the generated externs directly. Ideal scenario would be to have both options of course.

### Interoperability with C#

- It is also not a priority to make the generated code interoperable as is with regular C#. My main priority is to allow to perfectly run a Haxe project from a C# runtime (dotnet, Unity...).

- So at the moment, there is no such thing as `@:nativeGen` to make the output more "C# friendly". This, or a similar feature could be considered in the future, however.

### C# exported as source / text

Being able to export actual C# code (vs a DLL) is intended. It makes it easier to generate code that targets third party APIs without needing to reference the related DLLs. For instance, in the case of Unity, you can simply drop the output C# and let Unity Editor build it, and you can even use C# conditional compilation (in addition to the haxe one), whereas if you wanted to export a DLL that references Unity API, you'd need to figure out first how to link with the DLLs that provide the correct APIs, or write stubs that accurately replace it. And you can always make a DLL from that C# output if needed.

---

There are still a lot of things to do and review, before considering any of this ready to use : I need to make sure there is nothing stupid in the generated code that would destroy performances, catch any remaining edge case that may not be covered by the unit tests, and the target needs to be tested and profiled on actual use cases (like Ceramic's Unity export as mentioned earlier).

Regardless, I'm interested to have your feedback on this new C# target work. What do you think? I'm personally excited to finally be able to output C# from Haxe > 4 in the foreseeable future.

### CODE REVIEW

**Comment by Simn** [comment]
Oh my God... I know I've been joking that the community can just make their own C# target, but I didn't expect anyone to actually come through and do it... And it already passing the unit tests is a huge achievement, congrats on that!

There's obviously a lot I'll have to check here, but I'm not opposed to this on principle. The only immediate objection I have is that massive `system` package in std. This would have to go to some haxelib similar to what hxcs was for the old C# target.

I'll check this out properly some time next week and let you know my thoughts. From taking a quick look at gencs.ml it looks pretty good. And the changes to the rest of the compiler are minimal, which is a very good sign.

Edit: I would like to report a Windows newline problem!

```
utest/ui/text/PrintReport.hx:52:
                                 assertations: 11702
                                                    successes: 11702
                                                                    errors: 0
                                                                             failures: 0
                                                                                        warnings: 0
                                                                                                   execution time: 2

                                                                                                                    results: ALL TESTS OK (success: true)
```

**Comment by jeremyfa** [comment]
Ah, I'll have to test more on Windows for sure (been working on this from a mac mostly, but I'll summon the windows PC).

Regarding the `system` package, the rationale I had was to try to avoid depending on an external haxelib to export C# from Haxe, given that this `cs.system` package is needed if you use `sys` things in your haxe code, but if that's really a blocker, sure we can move that to a separate lib.

This PR is a draft anyway, so I don't consider it ready to be merged as is, feedback to improve it is definitely welcome, and I'll look at this windows newline thing!

File: src/compiler/args.ml [hunk 1]
```diff
@@ -93,6 +93,9 @@ let parse_args com =
 			actx.jvm_flag <- true;
 			set_platform com Jvm dir;
 		),"<file>","generate JVM bytecode into target file");
+		("Target",["--cs"],["-cs"],Arg.String (fun dir ->
```
Comment (kLabz) [inline src/compiler/args.ml:98]: Is this new C# target 100% compatible with haxe <= 4 C# target? If not, it will probably need a new "target name" I think :thinking:

File: std/haxe/ds/EnumValueMap.hx [hunk 2]
```diff
@@ -80,7 +80,7 @@ class EnumValueMap<K:EnumValue, V> extends haxe.ds.BalancedTree<K, V> implements
 
 	function compareArg(v1:Dynamic, v2:Dynamic):Int {
 		var vt1 = Type.typeof(v1);
-		var vt2 = Type.typeof(v1);
+		var vt2 = Type.typeof(v2);
```
Comment (kLabz) [inline std/haxe/ds/EnumValueMap.hx:83]: Wait what?

File: tests/sys/compile-fs.hxml [hunk 3]
```diff
@@ -1,4 +1 @@
-# comment the following line to disable testing the filesystem with invalid
-# Unicode codepoints; these will not work on APFS
 
--D TEST_INVALID_UNICODE_FS
```
Comment (kLabz) [inline tests/sys/compile-fs.hxml:4]: Why change this here?

File: src-json/meta.json
Comment (kLabz) [inline src-json/meta.json:1]: I expected more changes here and in define.json :thinking:

**Comment by kLabz** [comment]
Maybe tests removed in https://github.com/HaxeFoundation/haxe/pull/11551 should be reintroduced here?

File: src/compiler/args.ml (hunk 1 above)
Comment (Simn) [reply src/compiler/args.ml:98]: I think it's fine in a major change. Sure it might cause some confusing errors initially but people will survive.

File: tests/sys/compile-fs.hxml (hunk 3 above)
Comment (tobil4sk) [reply tests/sys/compile-fs.hxml:4]: Having to make this file empty is expected to be required locally on mac but yes, it shouldn't be checked into the repo

File: std/haxe/ds/EnumValueMap.hx (hunk 2 above)
Comment (kLabz) [reply std/haxe/ds/EnumValueMap.hx:83]: Eh.. nice catch but that fix came out of nowhere :sweat_smile: 

https://try.haxe.org/#7e1A0701

```haxe
function main() {
  var map = new Map<Foo, Bool>();
  map.set(Bar("ohno"), true);
  trace(map.exists(Bar(null))); // true
}

enum Foo {
    Bar(s:String);
}
```

File: src-json/meta.json
Comment (Simn) [reply src-json/meta.json:1]: Me too, I'm missing a `-D cs.net-ver` already because I don't have 8 installed!

File: src/compiler/args.ml (hunk 1 above)
Comment (jeremyfa) [reply src/compiler/args.ml:98]: Why would we need a new target name when there is no existing `cs` target in Haxe 5 anyway? If a codebase need to adapt for haxe 5's C# target it can already use haxe version conditional compilation, which is not worse than having `cs` that doesn't work at all on haxe 5 anyway.

File: std/haxe/ds/EnumValueMap.hx (hunk 2 above)
Comment (jeremyfa) [reply std/haxe/ds/EnumValueMap.hx:83]: Yes, this was already there I believe.

File: tests/sys/compile-fs.hxml (hunk 3 above)
Comment (jeremyfa) [reply tests/sys/compile-fs.hxml:4]: My bad, I think this slipped-in when it shouldn't have. I'll revert it and check how it goes.

**Comment by jeremyfa** [comment]
> Maybe tests removed in https://github.com/HaxeFoundation/haxe/pull/11551 should be reintroduced here?

Yes, definitely!

File: src-json/meta.json
Comment (jeremyfa) [reply src-json/meta.json:1]: Didn't handle any C# specific meta at the moment indeed, just added `@:unreflective` for now. Open to suggestions of course!

File: std/haxe/ds/EnumValueMap.hx (hunk 2 above)
Comment (jeremyfa) [reply std/haxe/ds/EnumValueMap.hx:83]: (would have been better on a separate commit yeah 😔)

Comment (kLabz) [reply std/haxe/ds/EnumValueMap.hx:83]: It's ok I'll add a test and that fix on development

**Comment by Simn** [comment]
Here's a code generation problem for you:

```haxe
function main() {
	while (true) {
		switch (Std.random(10)) {
			case 0:
				break;
			case 1:
				trace("1");
			case _:
		}
	}
}
```

Generates:

```cs
        public static void main()
        {
            unchecked
            {
                while ((true))
                    switch ((global::haxe.root.Std.random(10)))
                    {
                        case 0:
                            goto _hx_break_0;
                        case 1:
                            global::haxe.lang.Runtime.invokeFunction(global::haxe.Log.trace, global::haxe.root.Array.__ofDynLiteral(new object[] { "1", global::haxe.lang.HaxeDynamicObject._hx_create(global::haxe.root.Array.__ofDynLiteral(new object[] { "fileName", "source/Main.hx", "lineNumber", 7, "className", "_Main.Main_Fields_", "methodName", "main" })) }));
                            break;
                        default:
                            {
                            }
                            break;
                    }
                
            _hx_break_0:
                
            }
        }
```

```
C:\git\haxerepro\export\cs\_Main\Main_Fields_.cs(36,25): error CS1525: Invalid expression term '}' [C:\git\haxerepro\export\cs\Project.csproj]
C:\git\haxerepro\export\cs\_Main\Main_Fields_.cs(36,25): error CS1002: ; expected [C:\git\haxerepro\export\cs\Project.csproj]
```

**Comment by Simn** [comment]
And another one:

```haxe
function run<T>(f:() -> T):T {
	return f();
}

function main() {
	run(() -> {});
}
```

Generates:

```cs
        public static void main()
        {
            _ = ((void)(global::_Main.Main_Fields_.run(new _hx_Closure_Main_Fields__main_0())));
        }
```

```
C:\git\haxerepro\export\cs\_Main\Main_Fields_.cs(25,19): error CS1547: Keyword 'void' cannot be used in this context [C:\git\haxerepro\export\cs\Project.csproj]
```

This is one of these evil void-as-type-parameter cases.

**Comment by jeremyfa** [comment]
The first one should be fixed now: https://github.com/jeremyfa/haxe/commit/e5285dc8a5c45ff6966444510da3c3dd17a020ab

Surprised that `trace` is invoked dynamically, but maybe that's normal, because it can be replaced at runtime?

**Comment by Simn** [comment]
Yes, `haxe.Log.trace` is a `dynamic function`. This surprises everyone at some point.

File: tests/misc/cs/projects/Bootstrap/Main.hx [hunk 4]
```diff
@@ -1915,6 +1916,21 @@ class Main {
...
+			switch (Std.random(10)) {
+				case 0:
+					break;
+				case 1:
+					t(true);
+				case _:
+			}
+			break;
+		}
+		t(true);
```
Comment (Simn) [inline tests/misc/cs/projects/Bootstrap/Main.hx:1943]: This way it probably doesn't actually test it because the label was fine if there's a trailing statement, so it'd be better to put this before the loop. There's also `utest.Assert.pass()` for such situations.

Comment (jeremyfa) [reply tests/misc/cs/projects/Bootstrap/Main.hx:1943]: Right, I checked that the output is correct but you're right this makes the test useless 😅

**Comment by Simn** [comment]
Nice, I like how easy these are to fix.

At some point we'll have to get coroutines working too, which requires some generator support to expand argument types in the right places. This doesn't work at the moment:

```haxe
@:coroutine function f() {}

function main() {
	f(null);
}
```

Look how it's handled in other targets where they use use `follow_with_coro` and `expand_coro_type`.

For a real test case, you can checkout https://github.com/HaxeFoundation/hxcoro and then in the `tests` directory run `haxe build-base.hxml --cs bin/cs`. Compilation works fine (you need to merge development branch here because I fixed something there), but there's lots of errors related to `Corutine<>` which probably all have the same cause.

There's 82 occurrences of `TFun` in gencs, so... Good luck!

**Comment by pinchasn-prog** [comment]
👍For this PR, it's a lifesaver, GitHub is missing the option to star a PR....

File: tests/runci/targets/Cs.hx
Comment (jeremyfa) [inline tests/runci/targets/Cs.hx:1]: It's an "it works on my machine" moment. Hold on.

**Comment by jeremyfa** [comment]
https://github.com/jeremyfa/haxe/actions/runs/21834162316

Looks like the hxcoro tests are passing now, I added them to the C# test suite in CI

File: tests/sys/compile-fs.hxml (hunk 3 above)
Comment (kLabz) [reply tests/sys/compile-fs.hxml:4]: Would be better to properly revert all changes in this file (currently losing the comment for no reason)

Comment (jeremyfa) [reply tests/sys/compile-fs.hxml:4]: Not sure to understand what you mean, I did revert the change (with an additional commit). What else should be done about this?

Comment (kLabz) [reply tests/sys/compile-fs.hxml:4]: Github UI is barely usable on such a big PR, but current diff is like this (instead of no diff at all)
image

Comment (jeremyfa) [reply tests/sys/compile-fs.hxml:4]: Ah, that comment! Right, I didn't restore that, will do.

**Comment by jeremyfa** [comment]
Regarding the C# system externs package currently embedded in this PR, I think I'll explore another solution that doesn't require having those externs at all and rely on a similar system as haxe4's C# target (extracting externs on the fly from the dlls at build time), but I'll also take this as an opportunity to extract actual documentation from the XMLs that come along the DLLs instead of having the typings only, and I'll still make sure we can export plain C# externs from a DLL+XML manually if needed.

I'll likely fork `hxcs` repository to put those, keep the compatibility with haxe4 but add what's needed to support this new C# target for haxe5 as well.

**Comment by jeremyfa** [comment]
So here is what I'm doing now regarding C# .NET externs:

- They are being read from a DLL, I removed the haxe extern files

- If an XML file (with the docs) is provided alongside the DLL, it is used as well for documenting the types

- There is a caching system in place to prevent from parsing the xml docs everytime we need it

- Forked `hxcs` to provide typings and docs for .NET Standard 2.1 API

- If a DLL references a C# type that we don't have, a (Dynamic) stub is provided instead. This allows to use a DLL without having to pull all the other DLLs it depends on (this was a pain on previous C# target)

- The DLL typing extraction is using the same ilib utility as in Haxe 4

- `-net-lib` is back (but need to test it more)

- `hxcs` is actually not a requirement. The .NET API can be provided explicitly (a custom one, or from another lib), but using `hxcs` is the most straightforward "default" option.

A lot of this needs further testing and ajustments but I went back to the stage where all unit tests are passing after moving from haxe file externs to this DLL approach similar to Haxe 4's C# target, so that's not so bad!

**Comment by Simn** [comment]
So after doing `haxelib git hxcs https://github.com/jeremyfa/hxcs` I can compile and run the hxcoro tests successfully with `-lib hxcs`, nice!

One things I've noticed: I didn't clean my output directory and initially was getting `C:\git\hxcoro\tests\bin\cs\hxcoro\task\StartableCoroTask.cs(10,88): error CS0234: The type or namespace name 'Coroutine<>' does not exist in the namespace 'haxe.coro' (are you missing an assembly reference?)` again. This is from a file that I have deleted since originally compiling, which means that it's a leftover that then gets picked up by the `dotnet run`. I suppose that's the expected behavior, I just mention it here because it's a bit tricky.

I've also tested the silly thread benchmark I reported to HL [here](https://github.com/HaxeFoundation/hashlink/issues/866) and C# performs quite well:

```
source/Main.hx:30: single-threaded (2), 134
source/Main.hx:48: multi-threaded (2), 85
source/Main.hx:30: single-threaded (5), 320
source/Main.hx:48: multi-threaded (5), 291
source/Main.hx:30: single-threaded (10), 564
source/Main.hx:48: multi-threaded (10), 489
```

Not a huge difference but the numbers are lower, which means it's generally working right.

**Comment by Simn** [comment]
Checking some of the output of the hxcoro tests, one thing I really don't like is the huge number of `_hx_bind` calls at the entry point. This eagerly loads the entire application which comes with a substantial boot-up time. Since this seems to be about reflection support, the related data should be initialized much more lazily.

**Comment by tobil4sk** [comment]
> This is from a file that I have deleted since originally compiling, which means that it's a leftover that then gets picked up by the dotnet run.

Is this because the csproj file adds sources by entire directories? It should list files individually by name to avoid that. That's what hlc and hxcpp do, probably what old c# used to do as well?

**Comment by Simn** [comment]
> > This is from a file that I have deleted since originally compiling, which means that it's a leftover that then gets picked up by the dotnet run.
> 
> Is this because the csproj file adds sources by entire directories? It should list files individually by name to avoid that. That's what hlc and hxcpp do, probably what old c# used to do as well?

It actually doesn't list anything, so I guess that just just makes it pick up the entire directory.

**Comment by jeremyfa** [comment]
> Checking some of the output of the hxcoro tests, one thing I really don't like is the huge number of `_hx_bind` calls at the entry point. This eagerly loads the entire application which comes with a substantial boot-up time. Since this seems to be about reflection support, the related data should be initialized much more lazily.

That’s indeed for reflection. I don’t think I can avoid explicitly referencing the classes if I want to prevent aot trimming from discarding them (if not used statically), but you’re right that the actual data initialization involved could be more lazy (and not affect the startup time that much).

**Comment by jeremyfa** [comment]
So my last changes are making the _hx_bind calls more lazy.

I also addressed this file listing on .csproj, but instead of putting the file list directly inside the csproj, it is exported on a separate Sources.props file referenced by it. It makes it easier to replace or edit the csproj (manually or from a script) without having to deal with the file list itself.

**Comment by Simn** [comment]
`-D net-framework=net10.0` doesn't work, I think there are dash/underscore normalization problems. But you shouldn't use `raw_defined_value` anyway and add a proper define instead.

The `Sources.props` change looks good to make and makes sense.

I'm not sure what to make of all this reflection support code. It seems like the majority of the generated code is just there to support reflective stuff, which doesn't feel great. Isn't there a way to utilize C#'s own reflection capabilities to avoid/reduce all this cruft?

**Comment by jeremyfa** [comment]
> I'm not sure what to make of all this reflection support code. It seems like the majority of the generated code is just there to support reflective stuff, which doesn't feel great. Isn't there a way to utilize C#'s own reflection capabilities to avoid/reduce all this cruft?

The embedded data for supporting reflection on C# target is not new. The previous haxe 4 C# target also needed to embed similar data. The main difference is that it had a `FieldLookup.cs` class that maps hashes to names, which is then used at many places to avoid the string hashing when possible, but overall that was still the same idea: `__hx_getField()` with switches that map hashes to fields, closures and so on... If that data isn't there, it simply cannot work in scenario like Unity/AOT exporting C# to C++ via IL2CPP.

That doesn't mean we couldn't make things more configurable with some defines or metadata in the future to embed less info, for contexts that do have full C# reflection available, but I gave priority to the option that works in all contexts consistently, which also turns out to be the one option needed for the main use case I am personally having in practice.

> -D net-framework=net10.0 doesn't work, I think there are dash/underscore normalization problems. But you shouldn't use raw_defined_value anyway and add a proper define instead.

Makes sense. Will look at that.

**Comment by jeremyfa** [comment]
What I could check is if `someType.GetMembers()` (C# reflection) for `Type.getClassFields()` and `Type.getInstanceFields()` could work in AOT context too, instead of lazy initialized arrays for listing the fields, but I'm pretty sure `_hx_getField()` / `_hx_getStaticField()` with their switches for each field and function creation and caching boilerplate, will be required anyway.

**Comment by jeremyfa** [comment]
Well, I tried the `GetMembers()` solution (which is the one used by the previous C# target) for `Type.getClassFields()` and `Type.getInstanceFields()`. It works in JIT context, but it fails when using AOT: the member list is trimmed.

It would be possible to keep the members lists by annotating all haxe classes with `[DynamicallyAccessedMembers]`, but that would mean we are just asking the C# compiler to embed reflection data in the binary, possibly embed even more data than what we need instead of embedding it ourself explicitly, so I'm not really in favor of that.

Not against having a define that controls whether we export the C# for JIT contexts with full reflection available only (so less embedded boilerplate), or for all contexts (AOT included, with all embedded reflection code), but at least the latter is semantically correct in all cases, while the former is more in the _optimizations_ area in my opinion.

**Comment by Simn** [comment]
That already looks better now, thanks!

The next part that looks a little over-engineered is all this `_hx_invokeMethod` stuff. I understand what you're doing there and this index-based approach actually looks quite smart, but man is it a lot of generated code just to optimize field closures... Plus `_hx_closureCache` means that every single object has an increased memory footprint just to support this. I know it's not much because it's lazily initialized, but still, this seems like a lot of work for a rather rare case.

For what it's worth, I find it totally acceptable that `Reflect.field(obj, "someMethod")` allocates a new field closure object every time. We can still optimize statically known closures like `obj.someMethod` because in that case we _know_ that we need them, but assuming that every method might become a closure is a bit much in my opinion.

**Comment by jeremyfa** [comment]
Turns out this boilerplate for handling instance methods was written before I addressed the class method closures (static methods) resolution, which relies a bit more on delegates. I managed to remove this boilerplate for instance methods by using the same technique I use for class methods (except that I capture `this` in the delegate). That's definitely for the best.

Regarding the closure caching, I disabled it by default now, but I had to modify the `ClassMethodFunction` type so that equality comparison would work even if two identical closures are not the same object reference.

That caching can still be enabled with `-D cs.closure-cache`, which could still be useful in some situations, but it's now opt-in.

Now in the CI I test two scenarios:

- Export C# with default settings (no closure cache, no aot additional helpers for reflection) -> execute with JIT

- Export C# with `-D cs.aot -D cs.closure-cache`, compile using AOT and run the result

**Comment by Simn** [comment]
Looks better now!

> Regarding the closure caching, I disabled it by default now, but I had to modify the `ClassMethodFunction` type so that equality comparison would work even if two identical closures are not the same object reference.

Yeah, that's what I did for the JVM target too. It feels a bit like we're lying about the equality to satisfy some tests, but I guess it's not a real problem.

---

Next up, do we really need `_hx_emptyFactory` on every class? From what I'm seeing, it always calls a specific constructor, so it seems like that could just be done by `HaxeReflection.createEmpty` directly.

---

Also, try running the hxcoro tests with -dce no. I'm seeing this:

```
 5 | class CoroutineException extends Exception {}
   |       ^^^^^^^^^^^^^^^^^^
   | Could not patch constructor on this function because there isn't one
```

There's a chance that this is my fault, but it doesn't fail with genjvm so it might not be.

**Comment by tobil4sk** [comment]
> Yeah, that's what I did for the JVM target too. It feels a bit like we're lying about the equality to satisfy some tests, but I guess it's not a real problem.

I've seen code like this before, where the equality result would impact behaviour:
```haxe
addEventListener(MY_EVENT, this.method);
// ...
removeEventListener(MY_EVENT, this.method);
```

> 5 | class CoroutineException extends Exception {}
   |       ^^^^^^^^^^^^^^^^^^
   | Could not patch constructor on this function because there isn't one 

There is a similar issue on cppia, could be related:
https://github.com/HaxeFoundation/haxe/issues/12208#issuecomment-3314987116

**Comment by jeremyfa** [comment]
Not sure about that last change to address the -dce no with hxcoro, so I shared the analysis made by the coding assistant (which I used for this change) in the commit description.

Those factory constructor methods were not really needed indeed so I inlined them within _hx_bind.

**Comment by Simn** [comment]
I tried adding `@:keep function __shiftStack():Void { }` to JVM's `Exception` and that indeed reproduces the problem there too. I always forget what this actually does in practice, but I also don't see any point in adding a constructor just so we can shift its stack away, so I think the fix is good.

**Comment by Simn** [comment]
Regarding the factories, this is better but not quite what I meant because I'm trying to work towards the removal of `_hx_bind` as a whole, which I think should be possible. My uninformed approach to the constructor situation would be this in `creeateEmpty`:

```cs
try {
    SomehowLookUpConstructor(type).invoke((global::haxe.lang.EmptyConstructor)(null));
} catch(...) {
    // fallback
}
```

Or maybe `RuntimeHelpers.GetUninitializedObject` would just work for Haxe classes too? Not sure how that works.

And for `create`, having the factory function on the class seemed fine because it has to deal with all the argument type handling, so it would do something along the lines of:

```cs
try {
    SomehowLookUpStaticFunction(type, "_hx_factory").invoke(args);
} catch(...) {
    // fallback
}
```

**Comment by jeremyfa** [comment]
I don't see how we can avoid `_hx_bind()` in AOT trimming contexts, but now that we are at a stage where all the tests pass with that setup, we can try to remove it in JIT context and see if things are breaking or not 😬 

I'm aware of `RuntimeHelpers.GetUninitializedObject` but didn't try to use it yet, because the previous C# target was not using it either and was already relying on the `(EmptyContructor)null` technique, which I took as a hint that GetUninitializedObject is likely not behaving exactly the same, but we could take a second look at some point, maybe that's also something that would only fail in AOT contexts.

And my changes that replace the factory methods with delegates are not really changing how the arguments are handled because they are wrapped in `haxe.lang.Function`.

Sure we could put emptyFactory and factory functions resolution in the _hx_getStaticField embedded in the class instead of doing the wrapping at _hx_bind() stage eagerly however, that's a good change to make.

**Comment by tobil4sk** [comment]
@jeremyfa I wanted to test out the exception stack behaviour, so I made a small sample

```haxe
using haxe.CallStack;

class Test extends haxe.Exception {}

class TestChild extends Test {}

class TestGrandChild extends TestChild {
	public function new() {
		super("custom");
	}
}

function main() {
	trace(new TestGrandChild().stack.toString());
	trace(new TestChild("").stack.toString());
	trace(new Test("").stack.toString());
}
```

Dotnet fails to compile the c# code:
```
$ haxe --cs bin --main Main -lib hxcs
$ (cd bin && dotnet run)
/.../test-cs-exceptions/bin/_Main/Main_Fields_.cs(23,173): error CS1729: 'TestChild' does not contain a constructor that takes 3 arguments [/.../test-cs-exceptions/bin/Project.csproj]
/.../test-cs-exceptions/bin/haxe/root/TestGrandChild.cs(10,35): error CS1729: 'TestChild' does not contain a constructor that takes 3 arguments [/.../test-cs-exceptions/bin/Project.csproj]

The build failed. Fix the build errors and run again.
```
However, it looks like constructors like this are being generated, and it seems like they are missing the shiftStack call?
```c#
public Test(string message, global::haxe.Exception previous = default(global::haxe.Exception), object native = default(object)) : base(message, previous, native)
        {
        }
```

**Comment by jeremyfa** [comment]
Just a heads up to say that I'm currently busy on other matters so I'm not doing active work on this at the moment, but that's also because I'm waiting for the dust to settle a bit regarding Haxe 5.

I believe it will be a good time to resume active work on this when at least we have a proper release candidate of Haxe 5 or even the first stable version released, which make things simpler than trying to keep up in real time with the changes happening on the dev branch. That will also be a time when I'll more eagerly start integrating Haxe 5 for my bigger projects (Ceramic, which needs the new C# target etc...). Meanwhile there is absolutely no urgency to finish this right now, and I'm already very happy that we got where we are now.

Will definitely get back to this when the time is right!

**Comment by MTUnixic** [comment]
any updates on this?

**Comment by jeremyfa** [comment]
> any updates on this?

Same status as my last message. I'll wait for Haxe 5 to settle (likely when it is published as a first stable release) before resuming work on that, so it's just on hold for now :)

**Comment by MTUnixic** [comment]
i see, good luck on your project it looks awesome
